### PR TITLE
Use ApplicativeError instead of fromListUnsafe for NonEmptyList

### PR DESCRIPTION
**Pull request by tbrown1979**
I made it so the query parameter endpoints will handle the case where the `List` of params couldn't be turned into a `NonEmptyList`. In that case we use the `fromEither` to register a failure that we can then handle however we see fit - such as returning a `BadRequest`. 

Please let me know if there's any improvements or changes you'd like to see.

### CODE REVIEW

File: src/main/scala/io/github/pauljamescleary/petstore/endpoint/PetEndpoints.scala [hunk 1]
```diff
@@ -109,28 +112,28 @@ class PetEndpoints[F[_]: Effect] extends Http4sDsl[F] {
...
         // We have a list of valid statuses, find them and return
-        for {
-          retrieved <- petService.findByStatus(NonEmptyList.fromListUnsafe(statuses))
+        (for {
+          statusesNel <- AE.fromEither(statuses.toNel.toRight(EmptyStatuses))
+          retrieved <- petService.findByStatus(statusesNel)
           resp <- Ok(retrieved.asJson)
-        } yield resp
+        } yield resp).recoverWith {
+        case EmptyStatuses => BadRequest("tag parameter not specified")
```
Comment (pauljamescleary) [inline src/main/scala/io/github/pauljamescleary/petstore/endpoint/PetEndpoints.scala]: This would be "status parameter not specified"

Comment (pauljamescleary) [reply src/main/scala/io/github/pauljamescleary/petstore/endpoint/PetEndpoints.scala]: Would be better if it followed the "action"  in other cases...

```
val action = for {
  statusesNel <- AE.fromEither(statuses.toNel.toRight(EmptyStatuses))
  retrieved <- petService.findByStatus(statusesNel)
  resp <- Ok(retrieved.asJson)
} yield resp

action .flatMap { 
 ...
}
```

File: src/main/scala/io/github/pauljamescleary/petstore/endpoint/PetEndpoints.scala [hunk 2]
```diff
@@ -109,28 +112,28 @@ class PetEndpoints[F[_]: Effect] extends Http4sDsl[F] {
...
-      case GET -> Root / "pets" / "findByTags" :? TagMatcher(Valid(Nil)) =>
-        BadRequest("tag parameter not specified")
-
       case GET -> Root / "pets" / "findByTags" :? TagMatcher(Valid(tags)) =>
-        for {
-          retrieved <- petService.findByTag(NonEmptyList.fromListUnsafe(tags))
-          resp <- Ok(retrieved.asJson)
-        } yield resp
+        (for {
+          tagsNel   <- AE.fromEither(tags.toNel.toRight(EmptyTags))
```
Comment (pauljamescleary) [inline src/main/scala/io/github/pauljamescleary/petstore/endpoint/PetEndpoints.scala]: We do the same thing here twice.  

Honestly, I think this should come out of the box with http4s.  We need a required, mult query param decoder.

Perhaps we can implement our own "RequiredMultiQueryParamDecoderMatcher" and contribute it back to http4s?

Comment (tbrown1979) [reply src/main/scala/io/github/pauljamescleary/petstore/endpoint/PetEndpoints.scala]: That's a great idea! I'll look into that :)

**Pull request closed without merging**